[Oxidize DAGCircuit] Add: ancestors, descendants, bfs_successors to oxidized `DAGCircuit`

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
These commits add the `ancestors`, `descendants` and `bfs_successors` to the oxidiced `DAGCircuit`

### Details and comments
[WIP]

raynelfss committed:
Add: `ancestors`, `descendants`, `bfs_successors` to oxidized `DAGCircuit`

raynelfss committed:
Merge branch 'oxidize-dag' into oxidize-dag-family-tree

raynelfss committed:
Merge branch 'oxidize-dag' into oxidize-dag-family-tree

raynelfss committed:
Add: Rust-native methods
- Added rust native methods `ancestors`, `descendants` and `bfs_successors` that return iterators of node indices.

kevinhartman commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -43,6 +43,11 @@ use rustworkx_core::petgraph::prelude::StableDiGraph;
 use rustworkx_core::petgraph::stable_graph::{DefaultIx, IndexType, Neighbors, NodeIndex};
 use rustworkx_core::petgraph::visit::{IntoNodeReferences, NodeCount, NodeRef};
 use rustworkx_core::petgraph::Incoming;
+use rustworkx_core::traversal::{
```
Seeing this here, I think we should probably alias `rustworkx_core` as `rx` and then use fully qualified paths to make the callsites cleaner, i.e.

```rust
use rustworkx_core as rx;
```

And then at the call site, do `rx::traversal::ancestors(...)`.

No action needed, I can adjust after merge.

kevinhartman commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -2803,23 +2808,57 @@ def _format(operand):
+    #[pyo3(name="descendants")]
+    fn py_descendants(&self, py: Python, node: &DAGNode) -> PyResult<Py<PySet>> {
+        let descendants = self.descendants(node)
+            .filter_map(|node| self.get_node(py, node).ok());
```
The `get_node` method will only return an `Err` if something goes wrong when constructing the DAGNode, which really shouldn't happen. But, if it i does, we ought to propagate the error up through this method, rather than turning it into an `Option`.

I imagine the purpose of the `filter_map` you've got here is to ignore any nodes that aren't found. However, if the node ID you provide to `get_node` isn't found, it'll actually _panic_ and crash the process rather than return an `Err`.

Since we're calling `get_node` here using IDs that came from the graph itself, there is no risk of that happening, anyway 🙂.

I'd recommend doing something like:

```rust
let descendants: PyResult<Vec<_>> = self.descendants(node.node.unwrap).map(|node| self.get_node(py, node)).collect();
let descendant_set = PySet::new_bound(py, descendants?)?;
```

kevinhartman commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -3571,6 +3610,35 @@ impl DAGCircuit {
     }
 
+    /// Returns an iterator of the ancestors indices of a node.
+    pub fn ancestors<'a>(&'a self, node: &DAGNode) -> impl Iterator<Item = NodeIndex> + 'a {
```
For all of these Rust-native methods, we ought to deal in terms of `NodeIndex` for both the inputs and the outputs, i.e. you should change `node` to be a `NodeIndex`.

The reason being that `DAGNode` is really only a Python concept, since a Rust caller would only ever have a `DAGNode` if they were working with Python.

kevinhartman commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -3571,6 +3610,35 @@ impl DAGCircuit {
 
+    /// Returns an iterator of the ancestors indices of a node.
+    pub fn ancestors<'a>(&'a self, node: &DAGNode) -> impl Iterator<Item = NodeIndex> + 'a {
+        if let Some(node_index) = node.node {
```
I would just do `node.node.unwrap()` for now. Eventually, I will do a pass over everything in our joint PR to take care of a proper check using `#[pyo3(from_py_with = "...")]` (https://pyo3.rs/main/function#per-argument-options).

kevinhartman commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -2803,23 +2808,57 @@ def _format(operand):
+    fn py_bfs_successors(&self, py: Python, node: &DAGNode) -> PyResult<Py<PyList>> {
         // return iter(rx.bfs_successors(self._multi_graph, node._node_id))
-        todo!()
+        let successor_index = self.bfs_successors(node).filter_map(|(node, nodes)| {
```
Same comment here regarding not needing `filter_map` 🙂.

raynelfss committed:
Fix: Avoid checking validity of `NodeIndex`
- Make rust_native `ancestors`, `descendants` and `bfs_successors` return instances of `NodeIndex`.

raynelfss committed:
Fix: Filter out rust native functions to match previous behavior.
- The recently implemented rustwork-x core functions include the origin node and instances with no successors.
- Filter out those instances from the iterators.

raynelfss committed:
Format: Run cargo fmt

kevinhartman reviewed (approved):
LGTM, thanks for doing this!
